pkg/intoto/slsa_provenance/v1/provenance.go

```diff
@@ -0,0 +1,227 @@
+	DownloadLocation string `json:"downloadLocation,omitempty"`
+
+	// The MIME Type (i.e., media type) of the described resource.
+	MediaType string `json:"mediaType,omitempty"`
+
+	// The contents of the resource. This field is REQUIRED unless either uri
+	// or digest is set.
+	Content []byte `json:"content,omitempty"`
+
+	// This field MAY be used to provide additional information or metadata
+	// about the resource that may be useful to the consumer when evaluating
+	// the attestation against a policy.
+	Annotations map[string]interface{} `json:"annotations,omitempty"`
+}
+
+// Builder represents the transitive closure of all the entities that are, by
+// necessity, trusted to faithfully run the build and record the provenance.
+type Builder struct {
+	// URI indicating the transitive closure of the trusted builder.
+	ID string `json:"id"`
+
+	// Version numbers of components of the builder.
+	Version map[string]string `json:"version,omitempty"`
+
+	// Dependencies used by the orchestrator that are not run within the
+	// workload and that do not affect the build, but might affect the
+	// provenance generation or security guarantees.
+	BuilderDependencies []ResourceDescriptor `json:"builderDependencies,omitempty"`
+}
+
+type BuildMetadata struct {
+	// Identifies this particular build invocation, which can be useful for
+	// finding associated logs or other ad-hoc analysis. The exact meaning and
+	// format is defined by builder.id; by default it is treated as opaque and
+	// case-sensitive. The value SHOULD be globally unique.
+	InvocationID string `json:"invocationID,omitempty"`
+
+	// The timestamp of when the build started.
+	StartedOn *time.Time `json:"startedOn,omitempty"`
+
+	// The timestamp of when the build completed.
+	FinishedOn *time.Time `json:"finishedOn,omitempty"`
+}
+
+// DockerBasedExternalParameters is a representation of the top level inputs to
+// a container-based build.
+type DockerBasedExternalParameters struct {
+	// The source GitHub repo
+	Source ResourceDescriptor `json:"source"`
+
+	// The Docker builder image
+	BuilderImage ResourceDescriptor `json:"builderImage"`
+
+	// Path to a configuration file relative to the root of the repository.
+	ConfigPath string `json:"configPath"`
+
+	// Unpacked build config parameters
+	Config BuildConfig `json:"buildConfig"`
+}
+
+// BuildConfig is a collection of parameters to use for building the artifact
+// in a container-based build.
+type BuildConfig struct {
+	// The path, relative to the root of the git repository, where the artifact
+	// built by the `docker run` command is expected to be found.
+	ArtifactPath string `toml:"artifact_path"`
+
+	// Build command that is passed to `docker run`.
+	Command []string `toml:"command"`
+}
+
+// ParseContainerBasedSLSAv1Provenance parses the given object as a
+// ProvenancePredicate, with its BuildDefinition.ExternalParameters parsed into
+// an instance of DockerBasedExternalParameters. Returns an error if any of the
```

Comment:
Nit: You could drop "Returns an error if any of the conversions is unsuccessful". Or is there anything you want to emphasize which I am missing?

Reply:
We generally document the error case like this.